fdbserver/BlobManifest.actor.cpp

```diff
@@ -379,35 +382,42 @@ class BlobManifestLoader : public ReferenceCounted<BlobManifestLoader> {
 	ACTOR static Future<Standalone<BlobGranuleRestoreVersion>> getGranule(Transaction* tr, KeyRangeRef range) {
 		state Standalone<BlobGranuleRestoreVersion> granuleVersion;
 		state KeyRange historyKeyRange = blobGranuleHistoryKeyRangeFor(range);
-		// reverse lookup so that the first row is the newest version
-		state RangeResult results =
-		    wait(tr->getRange(historyKeyRange, GetRangeLimits::BYTE_LIMIT_UNLIMITED, Snapshot::True, Reverse::True));
-		for (KeyValueRef row : results) {
-			state KeyRange keyRange;
-			state Version version;
-			std::tie(keyRange, version) = decodeBlobGranuleHistoryKey(row.key);
-			Standalone<BlobGranuleHistoryValue> historyValue = decodeBlobGranuleHistoryValue(row.value);
-			state UID granuleID = historyValue.granuleID;
-
-			std::vector<GranuleFileVersion> files = wait(listGranuleFiles(tr, granuleID));
-			if (files.empty()) {
-				dprint("Granule {} doesn't have files for version {}\n", granuleID.toString(), version);
-				continue; // check previous version
-			}
-
-			granuleVersion.keyRange = KeyRangeRef(granuleVersion.arena(), keyRange);
-			granuleVersion.granuleID = granuleID;
-			granuleVersion.version = files.back().version;
-			granuleVersion.sizeInBytes = granuleSizeInBytes(files);
-
-			dprint("Granule {}: \n", granuleVersion.granuleID.toString());
-			dprint("  {} {} {}\n", keyRange.toString(), granuleVersion.version, granuleVersion.sizeInBytes);
-			for (auto& file : files) {
-				dprint("  File {}: {} bytes\n", file.filename, file.sizeInBytes);
+		loop {
+			try {
+				// reverse lookup so that the first row is the newest version
+				state RangeResult results = wait(
+				    tr->getRange(historyKeyRange, GetRangeLimits::BYTE_LIMIT_UNLIMITED, Snapshot::True, Reverse::True));
+				for (KeyValueRef row : results) {
+					state KeyRange keyRange;
+					state Version version;
+					std::tie(keyRange, version) = decodeBlobGranuleHistoryKey(row.key);
+					Standalone<BlobGranuleHistoryValue> historyValue = decodeBlobGranuleHistoryValue(row.value);
+					state UID granuleID = historyValue.granuleID;
+
+					std::vector<GranuleFileVersion> files = wait(listGranuleFiles(tr, granuleID));
+
+					granuleVersion.keyRange = KeyRangeRef(granuleVersion.arena(), keyRange);
+					granuleVersion.granuleID = granuleID;
+					if (files.empty()) {
```

Comment:
if current granule is empty, we can do time-travel read to parent granules. restore still can proceed